### ISSUE TITLE
Add testing regarding SparseAdam state_dicts

### ISSUE THREAD

**Issue by janeyx99** [pytorch/pytorch#116507]
### 🚀 The feature, motivation and pitch

SparseAdam, unlike the other optimizers, does not test any state_dict functionality. All other optims call _test_state_dict, but due to SparseAdam's only handling sparse gradients, it does not easily fall into the architecture that already exists.

We should consider giving it its own closure that would update parameters like in `rosenbrock` and creating special case tests for it.

### Alternatives

Leave SparseAdam state_dict handling untested

cc @vincentqb @jbschlosser @albanD @crcrpar @mruberry @ZainRizvi
